controllers/component/component_adapter_test.go

```diff
@@ -0,0 +1,147 @@
+				ComponentName:  "component-sample",
+				Application:    "application-sample",
+				ContainerImage: "",
+				Source: applicationapiv1alpha1.ComponentSource{
+					ComponentSourceUnion: applicationapiv1alpha1.ComponentSourceUnion{
+						GitSource: &applicationapiv1alpha1.GitSource{
+							URL: SampleRepoLink,
+						},
+					},
+				},
+			},
+			Status: applicationapiv1alpha1.ComponentStatus{
+				LastBuiltCommit: "",
+			},
+		}
+		Expect(k8sClient.Create(ctx, hasComp)).Should(Succeed())
+
+		hasComp2 = &applicationapiv1alpha1.Component{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "component-second-sample",
+				Namespace: "default",
+			},
+			Spec: applicationapiv1alpha1.ComponentSpec{
+				ComponentName:  "component-second-sample",
+				Application:    "application-sample",
+				ContainerImage: "",
+				Source: applicationapiv1alpha1.ComponentSource{
+					ComponentSourceUnion: applicationapiv1alpha1.ComponentSourceUnion{
+						GitSource: &applicationapiv1alpha1.GitSource{
+							URL: SampleRepoLink,
+						},
+					},
+				},
+			},
+		}
+		Expect(k8sClient.Create(ctx, hasComp2)).Should(Succeed())
+	})
+
+	AfterAll(func() {
+		err := k8sClient.Delete(ctx, hasApp)
+		Expect(err == nil || errors.IsNotFound(err)).To(BeTrue())
+		err = k8sClient.Delete(ctx, hasComp)
+		Expect(err == nil || errors.IsNotFound(err)).To(BeTrue())
+		err = k8sClient.Delete(ctx, hasComp2)
+		Expect(err == nil || errors.IsNotFound(err)).To(BeTrue())
+	})
+
+	It("can create a new Adapter instance", func() {
+		Expect(reflect.TypeOf(NewAdapter(hasComp, hasApp, logger, loader.NewMockLoader(), k8sClient, ctx))).To(Equal(reflect.TypeOf(&Adapter{})))
+	})
+	It("ensures removing a component will result in a new snapshot being created", func() {
+		buf := bytes.Buffer{}
+
+		log := helpers.IntegrationLogger{Logger: buflogr.NewWithBuffer(&buf)}
+		adapter = NewAdapter(hasComp, hasApp, log, loader.NewMockLoader(), k8sClient, ctx)
+		adapter.context = loader.GetMockedContext(ctx, []loader.MockData{
+			{
+				ContextKey: loader.ApplicationContextKey,
+				Resource:   hasApp,
+			},
+			{
+				ContextKey: loader.ComponentContextKey,
+				Resource:   hasComp,
+			},
+			{
+				ContextKey: loader.ComponentContextKey,
+				Resource:   hasComp2,
+			},
+			{
+				ContextKey: loader.ApplicationComponentsContextKey,
+				Resource:   []applicationapiv1alpha1.Component{*hasComp, *hasComp2},
+			},
+		})
+
+		applicationSnapshots, err := adapter.loader.GetAllSnapshots(k8sClient, adapter.context, hasApp)
+		Expect(err).To(BeNil())
+		Expect(*applicationSnapshots).To(BeNil())
+		now := metav1.NewTime(metav1.Now().Add(time.Second * 1))
+		hasComp.SetDeletionTimestamp(&now)
+
+		Eventually(func() bool {
+			result, err := adapter.EnsureComponentIsCleanedUp()
+			return !result.CancelRequest && err == nil
+		}, time.Second*20).Should(BeTrue())
+
+		applicationSnapshots, err = adapter.loader.GetAllSnapshots(k8sClient, adapter.context, hasApp)
```

Comment:
In this case case, since the control plane could be slow, it's best to have an Eventually block that waits until the list of Snapshots is populated, similar to how it's done [here](https://github.com/redhat-appstudio/integration-service/blob/272ccefcfc31504768a5e28780ee2fdb144b51e3/controllers/binding/snapshotenvironmentbinding_adapter_test.go#L372-L375) for integrationTestScenarios.